File: src/Sonirama.Api/Domain/Entities/Order.cs
using Sonirama.Api.Domain.Enums;

namespace Sonirama.Api.Domain.Entities;

public sealed class Order
{
    public Guid Id { get; set; } = Guid.NewGuid();
    public string Number { get; set; } = default!;
    public Guid UserId { get; set; }
    public User? User { get; set; }
    public OrderStatus Status { get; set; } = OrderStatus.PendingApproval;
    public decimal Subtotal { get; set; }
    public decimal DiscountTotal { get; set; }
    public decimal Total { get; set; }
    public string Currency { get; set; } = "ARS";
    public string? UserNotes { get; set; }
    public string? AdminNotes { get; set; }
    public string? RejectionReason { get; set; }
    public string? CancellationReason { get; set; }
    
    /// <summary>Reason for modification by admin</summary>
    public string? ModificationReason { get; set; }
    
    /// <summary>Admin who modified the order</summary>
    public Guid? ModifiedByUserId { get; set; }
    
    /// <summary>When the order was modified by admin</summary>
    public DateTime? ModifiedAtUtc { get; set; }
    
    /// <summary>Original total before modification (for comparison)</summary>
    public decimal? OriginalTotal { get; set; }
    
    public Guid? ApprovedByUserId { get; set; }
    public Guid? RejectedByUserId { get; set; }
    public Guid? ReadyByUserId { get; set; }
    public Guid? CompletedByUserId { get; set; }
    public DateTime CreatedAtUtc { get; set; } = DateTime.UtcNow;
    public DateTime UpdatedAtUtc { get; set; } = DateTime.UtcNow;
    public DateTime? ApprovedAtUtc { get; set; }
    public DateTime? RejectedAtUtc { get; set; }
    public DateTime? ConfirmedAtUtc { get; set; }
    public DateTime? ReadyAtUtc { get; set; }
    public DateTime? CompletedAtUtc { get; set; }
    public DateTime? CancelledAtUtc { get; set; }

    public ICollection<OrderItem> Items { get; set; } = new List<OrderItem>();
}

File: webapp/src/types/order.ts
/**
 * Order types
 */

export type OrderStatus =
  | 'PendingApproval'
  | 'Approved'
  | 'Rejected'
  | 'ModificationPending'
  | 'Confirmed'
  | 'ReadyForPickup'
  | 'Completed'
  | 'Cancelled';

export interface OrderItemDto {
  productId: string;
  productCode: string;
  productName: string;
  productImageUrl?: string | null;
  productImageAlt?: string | null;
  quantity: number;
  unitPrice: number;
  discountPercent: number;
  unitPriceWithDiscount: number;
  lineTotal: number;
  subtotal: number;
  product?: {
    images?: { url: string; altText: string; isPrimary: boolean }[];
  };
}

export interface OrderStatusHistoryDto {
  status: OrderStatus;
  changedAtUtc: string;
  changedByUserId?: string;
  notes?: string;
}

export interface OrderDto {
  id: string;
  number: string;
  orderNumber: string; // alias for number
  status: OrderStatus;
  userId: string;
  userPhoneNumber?: string | null;
  userDiscountPercent: number;
  subtotal: number;
  discountTotal: number;
  bulkDiscountAmount: number;
  total: number;
  currency: string;
  userNotes: string | null;
  adminNotes: string | null;
  rejectionReason: string | null;
  cancellationReason: string | null;
  modificationReason: string | null;
  originalTotal: number | null;
  modifiedAtUtc: string | null;
  createdAtUtc: string;
  updatedAtUtc: string | null;
  approvedAtUtc: string | null;
  confirmedAtUtc: string | null;
  readyAtUtc: string | null;
  items: OrderItemDto[];
  statusHistory?: OrderStatusHistoryDto[];
}

export interface OrderSummaryDto {
  id: string;
  number: string;
  status: OrderStatus;
  userId: string;
  total: number;
  currency: string;
  createdAtUtc: string;
  updatedAtUtc: string | null;
  itemCount: number;
}

export interface OrderListParams {
  page?: number;
  pageSize?: number;
  query?: string;
  status?: OrderStatus;
  createdFromUtc?: string;
  createdToUtc?: string;
  sortBy?: 'CreatedAt' | 'Number' | 'Status' | 'Total';
  sortDir?: 'ASC' | 'DESC';
  userId?: string;
}

export interface OrderConfirmRequest {
  note?: string;
}

export interface OrderCancelRequest {
  reason: string;
}

export interface OrderApproveRequest {
  adminNotes?: string;
}

export interface OrderRejectRequest {
  reason: string;
}

export interface OrderReadyRequest {
  readyNotes?: string;
}

export interface OrderCompleteRequest {
  completionNotes?: string;
}

export interface OrderModifyItemRequest {
  productId: string;
  newQuantity: number;
}

export interface OrderModifyRequest {
  reason: string;
  items: OrderModifyItemRequest[];
  adminNotes?: string;
}

export interface OrderAcceptModificationsRequest {
  note?: string;
}

export interface OrderRejectModificationsRequest {
  reason: string;
}

/**
 * Get human-readable order status label
 */
export function getOrderStatusLabel(status: OrderStatus): string {
  const labels: Record<OrderStatus, string> = {
    PendingApproval: 'Pendiente de aprobación',
    Approved: 'Aprobado',
    Rejected: 'Rechazado',
    ModificationPending: 'Modificación pendiente',
    Confirmed: 'Confirmado',
    ReadyForPickup: 'Listo para retiro',
    Completed: 'Completado',
    Cancelled: 'Cancelado',
  };
  return labels[status] || status;
}

/**
 * Get PrimeReact severity for order status
 */
export function getOrderStatusSeverity(
  status: OrderStatus
): 'success' | 'info' | 'warning' | 'danger' | 'secondary' | null {
  const severities: Record<OrderStatus, 'success' | 'info' | 'warning' | 'danger' | 'secondary'> = {
    PendingApproval: 'warning',
    Approved: 'info',
    Rejected: 'danger',
    ModificationPending: 'warning',
    Confirmed: 'info',
    ReadyForPickup: 'success',
    Completed: 'success',
    Cancelled: 'danger',
  };
  return severities[status] || null;
}

